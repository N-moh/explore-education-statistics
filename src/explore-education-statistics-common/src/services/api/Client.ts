import { AxiosInstance, AxiosPromise, AxiosRequestConfig } from 'axios';

class Client {
  public readonly api: AxiosInstance;

  public constructor(api: AxiosInstance) {
    this.api = api;
  }

  public get<T = unknown>(
    url: string,
    config?: AxiosRequestConfig,
  ): Promise<T> {
    return this.unboxResponse(this.api.get(url, config));
  }

  public post<T = unknown>(
    url: string,
    data?: unknown,
    config?: AxiosRequestConfig,
  ): Promise<T> {
    return this.unboxResponse(this.api.post(url, data, config));
  }

  public put<T = unknown>(
    url: string,
    data?: unknown,
    config?: AxiosRequestConfig,
  ): Promise<T> {
    return this.unboxResponse(this.api.put(url, data, config));
  }

  public patch<T = unknown>(
    url: string,
    data?: unknown,
    config?: AxiosRequestConfig,
  ): Promise<T> {
    return this.unboxResponse(this.api.patch(url, data, config));
  }

  public delete<T = unknown>(
    url: string,
    config?: AxiosRequestConfig,
  ): Promise<T> {
    return this.unboxResponse(this.api.delete(url, config));
  }

  private unboxResponse<T>(promise: AxiosPromise<T>) {
    return promise.then(({ data }) => data);
  }
}

export default Client;
