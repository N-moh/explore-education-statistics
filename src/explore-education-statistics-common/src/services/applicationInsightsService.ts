import { ApplicationInsights } from '@microsoft/applicationinsights-web';

export { ApplicationInsights };

const appInsights = new ApplicationInsights({
  config: {
    autoTrackPageVisitTime: true,
  },
});

export function initApplicationInsights(
  instrumentationKey: string,
): ApplicationInsights {
  if (instrumentationKey) {
    appInsights.config.instrumentationKey = instrumentationKey;
    appInsights.loadAppInsights();

    // eslint-disable-next-line no-console
    console.log('Application Insights initialised');
  }

  return appInsights;
}

/**
 * Escape hatch to get the Application Insights client
 * directly. Prefer to use context if possible.
 */
export function getApplicationInsights(): ApplicationInsights {
  return appInsights;
}
