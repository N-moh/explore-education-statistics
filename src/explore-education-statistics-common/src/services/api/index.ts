import axios from 'axios';
import { commaSeparated } from '../util/paramSerializers';
import Client from './Client';

interface BaseUrls {
  content: string;
  data: string;
  notification: string;
}

export const baseUrl: BaseUrls = {
  content: process.env.CONTENT_API_BASE_URL,
  data: process.env.DATA_API_BASE_URL || '/api/data',
  notification: process.env.NOTIFICATION_API_BASE_URL,
};

export const contentApi = new Client(
  axios.create({
    baseURL: baseUrl.content,
    paramsSerializer: commaSeparated,
  }),
);

export const dataApi = new Client(
  axios.create({
    baseURL: baseUrl.data,
    paramsSerializer: commaSeparated,
  }),
);

export const notificationApi = new Client(
  axios.create({
    baseURL: baseUrl.notification,
    paramsSerializer: commaSeparated,
  }),
);

const urlToApi: { [key in keyof BaseUrls]: Client } = {
  content: contentApi,
  data: dataApi,
  notification: notificationApi,
};

/**
 * Explicitly set base URLs for clients if unable to
 * rely on environment variables to correctly set them.
 */
export function setApiBaseUrls(baseUrls: BaseUrls) {
  Object.assign(baseUrl, baseUrls);

  Object.entries(urlToApi).forEach(([key, client]) => {
    // eslint-disable-next-line no-param-reassign
    client.axios.defaults.baseURL = baseUrls[key as keyof BaseUrls];
  });
}
