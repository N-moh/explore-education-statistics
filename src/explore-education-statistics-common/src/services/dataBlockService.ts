import { dataApi } from '@common/services/api';
import TimePeriod, { TimePeriodCode } from '@common/services/types/TimePeriod';
import { Dictionary } from '@common/types/util';
import { Feature, Geometry } from 'geojson';

export enum GeographicLevel {
  Establishment = 'Establishment',
  LocalAuthority = 'Local_Authority',
  LocalAuthorityDistrict = 'Local_Authority_District',
  LocalEnterprisePartnership = 'Local_Enterprise_Partnership',
  MATOrSponsor = 'MAT_Or_Sponsor',
  MayoralCombinedAuthorities = 'Mayoral_Combined_Authorities',
  National = 'National',
  OpportunityAreas = 'Opportunity_Areas',
  ParliamentaryConstituency = 'Parliamentary_Constituency',
  Provider = 'Provider',
  Regional = 'Regional',
  RSCRegion = 'RSC_Region',
  School = 'School',
  Ward = 'Ward',
}

export interface Country {
  country_code: string;
  country_name: string;
}

interface Region {
  region_code: string;
  region_name: string;
}

interface LocalAuthority {
  new_la_code: string;
  old_la_code: string;
  la_name: string;
}

interface LocalAuthorityDistrict {
  sch_lad_code: string;
  sch_lad_name: string;
}

export interface DataBlockLocation {
  country?: Country;
  region?: Region;
  localAuthority?: LocalAuthority;
  localAuthorityDistrict?: LocalAuthorityDistrict;
}

export interface Result {
  filters: Set<string>;
  location: DataBlockLocation;
  measures: {
    [key: string]: string;
  };
  timeIdentifier: TimePeriodCode;
  year: number;
}

export interface DataBlockData {
  publicationId: string;
  releaseId: string;
  subjectId: number;
  releaseDate: Date;
  geographicLevel: GeographicLevel;
  result: Result[];
}

// --- Meta data

interface LabelValueMetadata {
  label: string;
  value: string;
}

export interface LabelValueUnitMetadata extends LabelValueMetadata {
  unit: string;
}

interface OptionListMetadata<T extends LabelValueMetadata> {
  options: T[];
}

interface OptionMetadata extends OptionListMetadata<LabelValueMetadata> {
  label: string;
}

interface TimePeriodOptionMetadata {
  label: string;
  code: TimePeriodCode;
  year: number;
}

interface FilterMetadata {
  hint: string;
  legend: string;
  options: Dictionary<OptionMetadata>;
}

interface TimePeriodMetadata {
  hint: string;
  legend: string;
  options: TimePeriodOptionMetadata[];
}

interface LocationOptionsMetaData {
  hint: string;
  legend: string;
  options: LabelValueMetadata[];
}

interface IndicatorMetadata extends OptionListMetadata<LabelValueUnitMetadata> {
  label: string;
}

export interface ResponseMetaData {
  subject: {
    id: number;
    label: string;
  };
  filters: Dictionary<FilterMetadata>;
  indicators: Dictionary<IndicatorMetadata>;
  timePeriod: TimePeriodMetadata;
  locations: {
    LocalAuthority: LocationOptionsMetaData;
    LocalAuthorityDistrict: LocationOptionsMetaData;
    Regional: LocationOptionsMetaData;
    National: LocationOptionsMetaData;
  };
}

export interface DataBlockGeoJsonProperties {
  // these are what is required
  code: string;
  name: string;
  long: number;
  lat: number;

  // the following are just named here for easier finding in code completion and not required
  objectid?: number;
  ctry17cd?: string | null;
  ctry17nm?: string | null;
  lad17cd?: string | null;
  lad17nm?: string | null;

  // allow anything as this is an extension of the GeoJsonProperties object at its heart
  [name: string]: unknown;
}

export type DataBlockGeoJSON = Feature<Geometry, DataBlockGeoJsonProperties>;

export interface DataBlockLocationMetadata {
  value: string;
  label: string;
  geoJson: DataBlockGeoJSON[];
}

// ------------------------------------------

export interface DataBlockMetadata {
  indicators: Dictionary<LabelValueUnitMetadata>;
  filters: Dictionary<LabelValueMetadata>;
  timePeriod?: Dictionary<LabelValueMetadata>;
  timePeriods?: Dictionary<LabelValueMetadata>;
  locations?: Dictionary<DataBlockLocationMetadata>;
}

export interface DataBlockRequest {
  subjectId: number;
  geographicLevel: GeographicLevel;
  countries?: string[];
  localAuthorities?: string[];
  localAuthorityDistricts?: string[];
  regions?: string[];
  startYear: string;
  endYear: string;
  filters: string[];
  indicators: string[];
}

export interface DataBlockResponse {
  metaData: DataBlockMetadata;

  publicationId: string;
  releaseId: string;
  subjectId: number;
  releaseDate: Date;
  geographicLevel: GeographicLevel;
  result: Result[];
}

const DataBlockService = {
  buildTimePeriodMetadata(result: Result[]) {
    return result.reduce(
      (results: Dictionary<LabelValueMetadata>, { timeIdentifier, year }) => {
        const key = `${year}_${timeIdentifier}`;
        if (results[key]) return results;

        return {
          ...results,
          [key]: new TimePeriod(year, timeIdentifier),
        };
      },
      {},
    );
  },

  async getDataBlockForSubject(request: DataBlockRequest) {
    const response: DataBlockResponse = await dataApi.post('/Data', request);

    response.metaData.timePeriods =
      response.metaData.timePeriod ||
      DataBlockService.buildTimePeriodMetadata(response.result);

    return response;
  },
};

export default DataBlockService;
