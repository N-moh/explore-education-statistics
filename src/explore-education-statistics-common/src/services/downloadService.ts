import { contentApi } from '@common/services/api';
import downloadFile from '@common/utils/file/downloadFile';
import qs from 'qs';

export default {
  getFile(releaseId: string, fileId: string): Promise<Blob> {
    return contentApi.get<Blob>(`/releases/${releaseId}/files/${fileId}`, {
      responseType: 'blob',
    });
  },
  async downloadFiles(releaseId: string, fileIds: string[]): Promise<void> {
    const query = qs.stringify({ fileIds }, { arrayFormat: 'comma' });

    downloadFile(
      `${process.env.CONTENT_API_BASE_URL}/releases/${releaseId}/files?${query}`,
    );
  },
};
