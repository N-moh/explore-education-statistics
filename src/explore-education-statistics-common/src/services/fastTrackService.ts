import { dataApi } from '@common/services/api';
import { TableDataQuery } from '@common/services/tableBuilderService';
import { ConfiguredTable } from '@common/services/types/table';

export type FastTrackTable = ConfiguredTable & {
  query: TableDataQuery & {
    publicationId: string;
  };
};

export type FastTrackTableAndReleaseMeta = FastTrackTable & {
  releaseId: string;
  releaseSlug: string;
  latestData: boolean;
  latestReleaseTitle: string;
};

const fastTrackService = {
  getFastTrackTableAndReleaseMeta(
    id: string,
  ): Promise<FastTrackTableAndReleaseMeta> {
    return dataApi.get(`/fasttrack/${id}`);
  },
};

export default fastTrackService;
