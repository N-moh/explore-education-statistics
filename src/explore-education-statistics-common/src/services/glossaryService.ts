import { contentApi } from '@common/services/api';

export interface GlossaryEntry {
  title: string;
  slug: string;
  body: string;
}

export interface GlossaryCategory {
  heading: string;
  entries: GlossaryEntry[];
}

const glossaryService = {
  listGlossaryEntries(): Promise<GlossaryCategory[]> {
    return contentApi.get('/glossary', {});
  },
};

export default glossaryService;
