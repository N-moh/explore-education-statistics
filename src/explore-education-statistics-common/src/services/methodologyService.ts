import { Publication } from '@common/services/publicationService';
import { ContentBlock } from '@common/services/types/blocks';
import { contentApi } from './api';

export interface Methodology {
  id: string;
  title: string;
  published: string;
  lastUpdated: string;
  slug: string;
  publicationId: string;
  publication: Publication;
  content: {
    order: number;
    heading: string;
    caption: string;
    content: ContentBlock[];
  }[];
  annexes: {
    order: number;
    heading: string;
    caption: string;
    content: ContentBlock[];
  }[];
}

export default {
  getMethodology(methodologySlug: string): Promise<Methodology> {
    return contentApi.get(`/methodologies/${methodologySlug}`);
  },
};
