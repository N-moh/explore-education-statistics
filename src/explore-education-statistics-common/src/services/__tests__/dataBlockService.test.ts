/* eslint-disable prefer-destructuring */
import dataBlockService, {
  DataBlockRequest,
  GeographicLevel,
} from '@common/services/dataBlockService';
import testData from '@common/modules/find-statistics/components/charts/__tests__/__data__/testBlockData';

import { dataApi } from '@common/services/api';

jest.mock('@common/services/api', () => {
  return {
    dataApi: {
      post: jest.fn(),
    },
  };
});

describe('dataBlockService', () => {
  beforeEach(() => {
    // @ts-ignore
    dataApi.post.mockImplementation(() => {
      return Promise.resolve(testData.response);
    });
  });

  const dataBlockRequest: DataBlockRequest = {
    subjectId: 1,
    geographicLevel: GeographicLevel.Country,
    timePeriod: {
      startYear: '2014',
      startCode: 'HT6',
      endYear: '2015',
      endCode: 'HT6',
    },
    filters: ['1', '2'],
    indicators: ['23', '26', '28'],
  };

  test('calls API /Data endpoint', async () => {
    const result = await dataBlockService.getDataBlockForSubject(
      dataBlockRequest,
    );

    expect(dataApi.post).toBeCalledWith('/Data', dataBlockRequest);

    expect(result).not.toBeUndefined();
    if (result) {
      expect(result).toMatchSnapshot(result);
    }
  });
});
