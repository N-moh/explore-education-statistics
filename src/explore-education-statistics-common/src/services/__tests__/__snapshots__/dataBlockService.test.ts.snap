// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataBlockService calls API /Data endpoint 1`] = `
Object {
  "geographicLevel": "National",
  "metaData": Object {
    "filters": Object {
      "1": Object {
        "label": "All Schools",
        "value": "1",
      },
      "2": Object {
        "label": "All Pupils",
        "value": "1",
      },
    },
    "indicators": Object {
      "23": Object {
        "label": "Unauthorised absence rate",
        "unit": "%",
        "value": "23",
      },
      "26": Object {
        "label": "Overall absence rate",
        "unit": "%",
        "value": "26",
      },
      "28": Object {
        "label": "Authorised absence rate",
        "unit": "%",
        "value": "28",
      },
    },
    "locations": Object {
      "E92000001": Object {
        "geoJson": Array [
          Object {
            "geometry": Object {
              "coordinates": Array [
                Array [
                  Array [
                    Array [
                      400869.7932000002,
                      652744.4809000008,
                    ],
                    Array [
                      400910.5075000003,
                      652415.0365999993,
                    ],
                    Array [
                      400178.47229999956,
                      652575.4967,
                    ],
                  ],
                ],
              ],
              "type": "MultiPolygon",
            },
            "properties": Object {
              "code": "E92000001",
              "lat": 53.230099,
              "long": -2.07813,
              "name": "England",
              "objectid": 1,
            },
            "type": "Feature",
          },
        ],
        "label": "England",
        "value": "E92000001",
      },
      "S92000001": Object {
        "geoJson": Array [
          Object {
            "geometry": Object {
              "coordinates": Array [
                Array [
                  Array [
                    Array [
                      400869.7932000002,
                      652744.4809000008,
                    ],
                    Array [
                      400910.5075000003,
                      652415.0365999993,
                    ],
                    Array [
                      400178.47229999956,
                      652575.4967,
                    ],
                  ],
                ],
              ],
              "type": "MultiPolygon",
            },
            "properties": Object {
              "code": "S92000001",
              "lat": 53.230099,
              "long": -2.07813,
              "name": "Scotland",
              "objectid": 2,
            },
            "type": "Feature",
          },
        ],
        "label": "Scotland",
        "value": "S92000001",
      },
    },
    "timePeriods": Object {
      "2014_HT6": Object {
        "code": "HT6",
        "year": 2014,
      },
      "2015_HT6": Object {
        "code": "HT6",
        "year": 2015,
      },
    },
  },
  "publicationId": "test",
  "releaseDate": Object {},
  "releaseId": "1",
  "result": Array [
    Object {
      "filters": Array [
        "1",
        "2",
      ],
      "location": Object {
        "country": Object {
          "country_code": "E92000001",
          "country_name": "England",
        },
        "localAuthority": Object {
          "la_name": "",
          "new_la_code": "",
          "old_la_code": "",
        },
        "localAuthorityDistrict": Object {
          "sch_lad_code": "",
          "sch_lad_name": "",
        },
        "region": Object {
          "region_code": "",
          "region_name": "",
        },
      },
      "measures": Object {
        "23": "3",
        "26": "10",
        "28": "5",
      },
      "timeIdentifier": "HT6",
      "year": 2014,
    },
    Object {
      "filters": Array [
        "1",
        "2",
      ],
      "location": Object {
        "country": Object {
          "country_code": "E92000001",
          "country_name": "England",
        },
        "localAuthority": Object {
          "la_name": "",
          "new_la_code": "",
          "old_la_code": "",
        },
        "localAuthorityDistrict": Object {
          "sch_lad_code": "",
          "sch_lad_name": "",
        },
        "region": Object {
          "region_code": "",
          "region_name": "",
        },
      },
      "measures": Object {
        "23": "3",
        "26": "10",
        "28": "5",
      },
      "timeIdentifier": "HT6",
      "year": 2015,
    },
  ],
  "subjectId": 1,
}
`;
